agent/lib_composer.c

```diff
@@ -0,0 +1,233 @@
+/*
+ * Copyright 2022 New Relic Corporation. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "php_agent.h"
+#include "fw_hooks.h"
+#include "fw_support.h"
+#include "nr_txn.h"
+#include "util_logging.h"
+#include "util_memory.h"
+#include "util_syscalls.h"
+
+static bool nr_execute_handle_autoload_composer_is_initialized() {
+  zend_class_entry* zce = NULL;
+
+  if (NULL == (zce = nr_php_find_class("composer\\installedversions"))) {
+    nrl_verbosedebug(NRL_INSTRUMENT,
+                     "Composer\\InstalledVersions class not found");
+    return false;
+  };
+
+  // the class is found - there's hope!
+  if (NULL == nr_php_find_class_method(zce, "getallrawdata")
+      || NULL == nr_php_find_class_method(zce, "getrootpackage")) {
+    nrl_verbosedebug(
+        NRL_INSTRUMENT,
+        "Composer\\InstalledVersions class found, but methods not found");
+    return false;
+  }
+
+  return true;
+}
+
+static int nr_execute_handle_autoload_composer_init(const char* vendor_path) {
+  char* code = NULL;
+  zval retval;
+  int result = -1;
+
+  if (nr_execute_handle_autoload_composer_is_initialized()) {
+    nrl_verbosedebug(NRL_INSTRUMENT, "%s: already initialized", __func__);
+    return NR_SUCCESS;
+  }
+
+  code = nr_formatf("include_once '%s/composer/InstalledVersions.php';",
+                    vendor_path);
+
+  result = zend_eval_string(code, &retval, "newrelic\\init_composer_api");
+  if (result != SUCCESS) {
+    nrl_verbosedebug(NRL_INSTRUMENT,
+                     "%s: zend_eval_string(%s) failed, result=%d", __func__,
+                     code, result);
+    nr_free(code);
+    return NR_FAILURE;
+  }
+
+  zval_dtor(&retval);
+  nr_free(code);
+
+  // Make sure runtime API is available after loading
+  // Composer\\InstalledVersions class:
+  if (!nr_execute_handle_autoload_composer_is_initialized()) {
+    nrl_verbosedebug(NRL_INSTRUMENT,
+                     "%s: unable to initialize Composer runtime API", __func__);
+    return NR_FAILURE;
+  }
+
+  return NR_SUCCESS;
+}
+
+static void nr_execute_handle_autoload_composer_get_packages_information(
+    const char* vendor_path) {
+  zval retval;  // This is used as a return value for zend_eval_string.
+                // It will only be set if the result of the eval is SUCCESS.
+  int result = -1;
+
+  // nurunlikely because this should alredy be ensured by the caller
+  if (nrunlikely(!NRINI(vulnerability_management_package_detection_enabled))) {
+    // do nothing when collecting package information for vulnerability
+    // management is disabled
+    return;
+  }
+
+  // nurunlikely because this should alredy be ensured by the caller
+  if (nrunlikely(!NRINI(vulnerability_management_composer_api_enabled))) {
+    // do nothing when use of composer to collect package info is disabled
+    return;
+  }
+
+  // clang-format off
+  char* getallrawdata
```

Comment:
Should this be a `const char*` since it doesn't seem to get modified after definition?
```suggestion
  const char* getallrawdata
```

Reply:
This actually had to be reverted because `zend_eval_string` expects `char *` not `const char *`, so making `code` a `const char *` does not make much sense because it would have to be casted back to `char *`.

Reply:
Are you sure? According to [php-src](https://github.com/php/php-src/blob/master/Zend/zend_execute_API.c#L1383) `zend_eval_string` expects a `const char *` for it's first argument, not `char *`

Reply:
Yes, I'm sure - https://github.com/newrelic/newrelic-php-agent/actions/runs/11016894669/job/30593507308#step:7:197:
```
/usr/local/src/newrelic-php-agent/agent/lib_composer.c: In function 'nr_execute_handle_autoload_composer_get_packages_information':
/usr/local/src/newrelic-php-agent/agent/lib_composer.c:126:26: error: passing argument 1 of 'zend_eval_string' discards 'const' qualifier from pointer target type [-Werror=discarded-qualifiers]
       = zend_eval_string(getallrawdata, &retval, "composer_getallrawdata.php");
                          ^~~~~~~~~~~~~
In file included from /usr/local/include/php/Zend/zend_API.h:31,
                 from /usr/local/src/newrelic-php-agent/agent/php_includes.h:18,
                 from /usr/local/src/newrelic-php-agent/agent/php_agent.h:38,
                 from /usr/local/src/newrelic-php-agent/agent/lib_composer.c:6:
/usr/local/include/php/Zend/zend_execute.h:49:37: note: expected 'char *' but argument is of type 'const char *'
 ZEND_API int zend_eval_string(char *str, zval *retval_ptr, char *string_name);
                               ~~~~~~^~~
```

Reply:
For future reference: PHP 7.2 - 7.4 do not have the `const` modifier. `const` was added to this function arg in PHP 8.0.